x/icactl/keeper/keeper.go

```diff
@@ -0,0 +1,118 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	capabilitykeeper "github.com/cosmos/cosmos-sdk/x/capability/keeper"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	icacontrollerkeeper "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts/controller/keeper"
+	icatypes "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v3/modules/core/24-host"
+	"github.com/crypto-org-chain/cronos/x/icactl/types"
+	"github.com/tendermint/tendermint/libs/log"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type (
+	Keeper struct {
+		cdc        codec.BinaryCodec
+		paramStore paramtypes.Subspace
+
+		icaControllerKeeper icacontrollerkeeper.Keeper
+		scopedKeeper        capabilitykeeper.ScopedKeeper
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	paramStore paramtypes.Subspace,
+	icaControllerKeeper icacontrollerkeeper.Keeper,
+	scopedKeeper capabilitykeeper.ScopedKeeper,
+) *Keeper {
+	// set KeyTable if it has not already been set
+	if !paramStore.HasKeyTable() {
+		paramStore = paramStore.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		cdc:        cdc,
+		paramStore: paramStore,
+
+		icaControllerKeeper: icaControllerKeeper,
+		scopedKeeper:        scopedKeeper,
+	}
+}
+
+// DoSubmitTx submits a transaction to the host chain on behalf of interchain account
+func (k *Keeper) DoSubmitTx(ctx sdk.Context, connectionID, owner string, msgs []sdk.Msg, timeoutDuration time.Duration) error {
+	portID, err := icatypes.NewControllerPortID(owner)
+	if err != nil {
+		return err
+	}
+
+	channelID, found := k.icaControllerKeeper.GetActiveChannelID(ctx, connectionID, portID)
+	if !found {
+		return sdkerrors.Wrapf(icatypes.ErrActiveChannelNotFound, "failed to retrieve active channel for port %s", portID)
+	}
+
+	channelCapability, found := k.scopedKeeper.GetCapability(ctx, host.ChannelCapabilityPath(portID, channelID))
+	if !found {
+		return sdkerrors.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
+	}
+
+	data, err := icatypes.SerializeCosmosTx(k.cdc, msgs)
+	if err != nil {
+		return err
+	}
+
+	packetData := icatypes.InterchainAccountPacketData{
+		Type: icatypes.EXECUTE_TX,
+		Data: data,
+	}
+
+	timeoutTimestamp := ctx.BlockTime().Add(timeoutDuration).UnixNano()
```

Comment:
is the arbitrary user-set timeout desirable? I recall there was one issue that for ICA, the channel will be closed after the timeout

Reply:
oh, this usually should go via the other functions that has `k.MinTimeoutDuration` ... maybe good to highlight that

Reply:
sure, added a comment.